File: content/tracks/dsa/problems/lc-0150-evaluate-reverse-polish-notation/Solution.java
import java.util.*;

class Solution {
    private static final Set<String> OPERATORS = Set.of("+", "-", "*", "/");

    public int evalRPN(String[] tokens) {
        Deque<Integer> stack = new ArrayDeque<>();
        for (String token : tokens) {
            if (!OPERATORS.contains(token)) {
                stack.push(Integer.parseInt(token));
                continue;
            }
            // The first value popped is the right operand.
            int right = stack.pop();
            int left = stack.pop();
            // Java's integer division already truncates toward zero.
            stack.push(switch (token) {
                case "+" -> left + right;
                case "-" -> left - right;
                case "*" -> left * right;
                default -> left / right;
            });
        }
        return stack.peek();
    }
}

File: content/tracks/dsa/problems/lc-0150-evaluate-reverse-polish-notation/solution.go
package main

import "strconv"

func evalRPN(tokens []string) int {
	stack := []int{}
	for _, token := range tokens {
		if token != "+" && token != "-" && token != "*" && token != "/" {
			value, _ := strconv.Atoi(token) // tokens are valid integers
			stack = append(stack, value)
			continue
		}
		// The first value popped is the right operand.
		right := stack[len(stack)-1]
		left := stack[len(stack)-2]
		stack = stack[:len(stack)-2]
		// Go's integer division already truncates toward zero.
		switch token {
		case "+":
			stack = append(stack, left+right)
		case "-":
			stack = append(stack, left-right)
		case "*":
			stack = append(stack, left*right)
		default:
			stack = append(stack, left/right)
		}
	}
	return stack[len(stack)-1]
}
